typescript/sdk/src/zksync/ZKDeployer.ts

```diff
@@ -0,0 +1,193 @@
+  }
+
+  public loadArtifact(contractTitle: string): Promise<any> {
+    const artifact = (zksyncArtifacts as ZkSyncArtifact[]).find(
+      ({ contractName, sourceName }) => {
+        if (contractName === contractTitle) {
+          return true;
+        }
+
+        const qualifiedName = `${sourceName}:${contractName}`;
+        if (contractTitle === qualifiedName) {
+          return true;
+        }
+
+        return false;
+      },
+    );
+
+    if (!artifact) {
+      throw new Error(
+        `No ZKSync artifact for contract ${contractTitle} found!`,
+      );
+    }
+
+    return artifact as any;
+  }
+
+  public static loadArtifactByBytecode(bytecodeExt: string): Promise<any> {
+    const artifact = (zksyncArtifacts as ZkSyncArtifact[]).find(
+      ({ bytecode }) => bytecode === bytecodeExt,
+    );
+
+    return artifact as any;
+  }
+
+  /**
+   * Estimates the price of calling a deploy transaction in ETH.
+   *
+   * @param artifact The previously loaded artifact object.
+   * @param constructorArguments List of arguments to be passed to the contract constructor.
+   *
+   * @returns Calculated fee in ETH wei
+   */
+  public async estimateDeployFee(
+    artifact: ZkSyncArtifact,
+    constructorArguments: any[],
+  ): Promise<ethers.BigNumber> {
+    const gas = await this.estimateDeployGas(artifact, constructorArguments);
+    const gasPrice = await this.zkWallet.provider.getGasPrice();
+    return gas.mul(gasPrice);
+  }
+
+  /**
+   * Estimates the amount of gas needed to execute a deploy transaction.
+   *
+   * @param artifact The previously loaded artifact object.
+   * @param constructorArguments List of arguments to be passed to the contract constructor.
+   *
+   * @returns Calculated amount of gas.
+   */
+  public async estimateDeployGas(
+    artifact: ZkSyncArtifact,
+    constructorArguments: any[],
+  ): Promise<ethers.BigNumber> {
+    const factoryDeps = await this.extractFactoryDeps(artifact);
+
+    const factory = new zk.ContractFactory(
+      artifact.abi,
+      artifact.bytecode,
+      this.zkWallet,
+      this.deploymentType,
+    );
+
+    // Encode deploy transaction so it can be estimated.
+    const deployTx = factory.getDeployTransaction(...constructorArguments, {
+      customData: {
+        factoryDeps,
+      },
+    });
+    deployTx.from = this.zkWallet.address;
+
+    return await this.zkWallet.provider.estimateGas(deployTx);
+  }
+
+  /**
+   * Sends a deploy transaction to the zkSync network.
+   * For now, it will use defaults for the transaction parameters:
+   * - fee amount is requested automatically from the zkSync server.
+   *
+   * @param artifact The previously loaded artifact object.
+   * @param constructorArguments List of arguments to be passed to the contract constructor.
+   * @param overrides Optional object with additional deploy transaction parameters.
+   * @param additionalFactoryDeps Additional contract bytecodes to be added to the factory dependencies list.
```

Comment:
I dont understand this
can you link me to some docs?

Reply:
Basically on zksync all bytecode must be known before deployment. That is the reason why MetaProxy and Cloning don't work on zksync.

Here is the explanation https://docs.zksync.io/build/developer-reference/ethereum-differences/contract-deployment